### PR TITLE
Fix exec in proxy

### LINKED ISSUE #2204: Websocket requests are not proxied correctly in Ansible-based operators
## Bug Report

**What did you do?**

I am trying to an Ansible `k8s_exec` module, which allows running the equivalent of `kubectl exec` commands to exec a command on a Pod via Ansible through the Python Kubernetes library. This allows me to write a task like:

```yaml
- name: Test a simple command.
  k8s_exec:
    namespace: '{{ meta.namespace }}'
    pod: '{{ tower_pod_name }}'
    command: date
```

Instead of installing `kubectl` on my operator image (added `COPY --from=lachlanevenson/k8s-kubectl:v1.16.2 /usr/local/bin/kubectl /usr/local/bin/kubectl` to my `build/Dockerfile`) and writing a task like:

```yaml
- name: Test kubectl exec.
  command: >
    kubectl exec -n {{ meta.namespace }} {{ tower_pod_name }} date
```

**What did you expect to see?**

When I run the same task as above on my system Ansible against a Kubernetes cluster, or even inside of the operator Pod's ansible container using `ansible-playbook` to run it, it executes successfully and registers the result of the command that was executed.

**What did you see instead? Under which circumstances?**

When it is run via the operator/`ansible-runner` using the operator's proxy, it results in the following error:

```
kubernetes.client.rest.ApiException: (0)
Reason: Handshake status 200 OK
```

It should be getting a `101` response from the Kubernetes API websocket.

Full error message from the failed task:

```
TASK [tower : Test a simple command.] ******************************************
task path: /opt/ansible/roles/tower/tasks/main.yml:38
 line 136, in <module>\n  File \"/tmp/ansible_k8s_exec_payload_x8PPAd/__main__.py\", line 123, in main\n  File \"/usr/lib/python2.7/site-packages/kubernetes/stream/stream.py\", line 32, in stream\n    return func(*args, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/kubernetes/client/apis/core_v1_api.py\", line 835, in connect_get_namespaced_pod_exec\n    (data) = self.connect_get_namespaced_pod_exec_with_http_info(name, namespace, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/kubernetes/client/apis/core_v1_api.py\", line 935, in connect_get_namespaced_pod_exec_with_http_info\n    collection_formats=collection_formats)\n  File \"/usr/lib/python2.7/site-packages/kubernetes/client/api_client.py\", line 321, in call_api\n    _return_http_data_only, collection_formats, _preload_content, _request_timeout)\n  File \"/usr/lib/python2.7/site-packages/kubernetes/client/api_client.py\", line 155, in __call_api\n    _request_timeout=_request_timeout)\n  File \"/usr/lib/python2.7/site-packages/kubernetes/str
eam/stream.py\", line 27, in _intercept_request_call\n    return ws_client.websocket_call(config, *args, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/kubernetes/stream/ws_client.py\", line 255, in websocket_call\n    raise ApiException(status=0, reason=str(e))\nkubernetes.client.rest.ApiException: (0)\nReason: Handshake status 200 OK\n\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

**Environment**
* operator-sdk version:

v0.11.0

* go version:

N/A

* Kubernetes version information:

1.16.2

* Kubernetes cluster kind: 

Molecule

* Are you writing your operator in ansible, helm, or go?

ansible

**Possible Solution**

N/A

**Additional context**

Relates to: https://github.com/geerlingguy/tower-operator/issues/5

### PR DESCRIPTION
**Pull request by fabianvf**
**Description of the change:**
Allows exec requests to go through the proxy without error

**Motivation for the change:**
It was impossible to make exec requests from Ansible code without circumventing the proxy

fixes #2204

### CODE REVIEW

File: pkg/ansible/proxy/proxy.go [hunk 1]
```diff
@@ -44,6 +44,8 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
+const AutoSkipCacheRE = "^/api/.*/pods/.*/exec,^/api/.*/pods/.*/attach"
```
Comment (estroz) [inline pkg/ansible/proxy/proxy.go]:
```suggestion
const AutoSkipCacheREList = "^/api/.*/pods/.*/exec,^/api/.*/pods/.*/attach"
```

File: pkg/ansible/proxy/proxy.go [hunk 2]
```diff
@@ -158,6 +160,10 @@ func Run(done chan error, o Options) error {
 		server.Handler = RequestLogHandler(server.Handler)
 	}
 	if !o.DisableCache {
+		autoSkipCacheRegexp, err := MakeRegexpArray(AutoSkipCacheRE)
```
Comment (estroz) [inline pkg/ansible/proxy/proxy.go]:
```suggestion
		autoSkipCacheRegexp, err := MakeRegexpArray(AutoSkipCacheREList)
```

File: pkg/ansible/proxy/proxy.go (hunk 1 above)
Comment (estroz) [reply pkg/ansible/proxy/proxy.go]: Can these be pre-compiled?

Comment (fabianvf) [reply pkg/ansible/proxy/proxy.go]: currently they're only compiled once, when the proxy starts, I don't think there's a safe way to compile them into a const though, compiling the regex either returns `regexp, err` or panics AFAICT

**Comment by fabianvf** [comment]
@estroz no security concerns as far as I'm aware, obviously the user can execute arbitrary code in pods but as operator authors I think they should be allowed to (and could in Golang or any other language as well). You can still restrict the operator by not granting it `pods/exec` permission in the `Role`

File: pkg/ansible/proxy/proxy.go (hunk 1 above)
Comment (estroz) [reply pkg/ansible/proxy/proxy.go]: Since its a list you'd have to do some annoying stuff with an `init` function, so the way you have it now is fine.

**Comment by fabianvf** [comment]
test failures unrelated, just the 429 issue we've been seeing

File: Makefile [hunk 3]
```diff
@@ -202,7 +202,7 @@ test-sanity: tidy build/operator-sdk lint
 
 TEST_PKGS:=$(shell go list ./... | grep -v -E 'github.com/operator-framework/operator-sdk/(hack/|test/)')
 test-unit: ## Run the unit tests
-	$(Q)go test -coverprofile=coverage.out -covermode=count -count=1 -short ${TEST_PKGS}
+	$(Q)go test -coverprofile=coverage.out -covermode=count -count=1 ${TEST_PKGS}
```
Comment (fabianvf) [inline Makefile]: Apparently we've never actually run any tests on the proxy because they're skipped on a short run :grimacing:

Comment (estroz) [reply Makefile]: Can you make a separate PR removing `-short` since this affects all unit tests?

**Comment by geerlingguy** [comment]
Also I can confirm that removing the `exec` permission kills the ability for `k8s_exec` to do anything. Right now I'm using `kubectl` installed in an Operator image to work around the proxy, but this PR would be ideal to allow me to do everything in Ansible via the API.

File: Makefile (hunk 3 above)
Comment (fabianvf) [reply Makefile]: :+1: https://github.com/operator-framework/operator-sdk/pull/2718

**Comment by fabianvf** [comment]
build is passing, markdown just failed due to 429 flake

**Pull request merged**